panel/widgets/chat.py

```diff
@@ -0,0 +1,1593 @@
+        self._composite.hide_header = not self.header
+
+    def _replace_placeholder(self, entry: ChatEntry | None = None) -> None:
+        """
+        Replace the placeholder from the chat log with the entry
+        if placeholder, otherwise simply append the entry.
+        Replacing helps lessen the chat log jumping around.
+        """
+        index = None
+        if self.placeholder_threshold > 0:
+            try:
+                index = self.value.index(self._placeholder)
+            except ValueError:
+                pass
+
+        if index is not None:
+            if entry is not None:
+                self._chat_log[index] = entry
+            elif entry is None:
+                self._chat_log.remove(self._placeholder)
+        elif entry is not None:
+            self._chat_log.append(entry)
+
+    def _build_entry(
+        self,
+        value: ChatEntry |  dict | Any,
+        user: str | None = None,
+        avatar: str | BinaryIO | None = None,
+    ) -> ChatEntry | None:
+        """
+        Builds a ChatEntry from the value.
+        """
+        if value is None:
+            return None
+
+        if not isinstance(value, (ChatEntry, dict)):
+            value = {"value": value}
+
+        if isinstance(value, dict):
+            if "value" not in value:
+                raise ValueError(
+                    f"If 'value' is a dict, it must contain a 'value' key, "
+                    f"e.g. {{'value': 'Hello World'}}; got {value!r}"
+                )
+            if user:
+                value.update(user=user)
+            if avatar:
+                value.update(avatar=avatar)
+            if self.width:
+                entry_params = {"width": int(self.width - 80), **self.entry_params}
+            else:
+                entry_params = self.entry_params
+            entry_params.update(renderers=self.renderers)
+            input_params = {**value, **entry_params}
+            entry = ChatEntry(**input_params)
+        else:
+            # must update one at a time so avatar is not overwritten
+            # by the default avatar
+            if user:
+                value.user = user
+            if avatar:
+                value.avatar = avatar
+            entry = value
+        return entry
+
+    def _upsert_entry(self, value: Any, entry: ChatEntry | None = None) -> ChatEntry | None:
+        """
+        Replace the placeholder entry with the response or update
+        the entry's value with the response.
+        """
+        if isinstance(value, str) and entry:
+            entry.value = value
+            return entry
+
+        user = self.callback_user
+        avatar = None
+        if isinstance(value, ChatEntry):
+            user = value.user
+            avatar = value.avatar
+        elif isinstance(value, dict):
+            user = value.get("user", user)
+            avatar = value.get("avatar", avatar)
+
+        updated_entry = self._build_entry(value, user=user, avatar=avatar)
```

Comment:
Seems like this may error if `value` is `None` because `updated_entry` will also be None.

Reply:
value should never be None I think.